### ISSUE TITLE
BinaryMatthewsCorrCoef outputs 0 if target and prediction contains only one case either positive or negative case

### ISSUE THREAD

**Issue by lokesh1199** [Lightning-AI/torchmetrics#1811]
### Discussed in https://github.com/Lightning-AI/torchmetrics/discussions/1810

<div type='discussions-op-text'>

<sup>Originally posted by **lokesh1199** May 25, 2023</sup>
Hi,

When the given target and pred contains only one case either positive or negative (case  - 1), The output of BinaryMatthewsCorrCoef is always 0. But we can clearly say that predictions are same as the targets. And for the case - 2, If the preds and targets are completely different, the MCC is also zero. Is this expected behaviour with MCC or some other issue.

```python
mcc = BinaryMatthewsCorrCoef()

# preds and targets are equal (case - 1)
print(mcc(torch.ones(10), torch.ones(10)))
print(mcc(torch.zeros(10), torch.zeros(10)))

# preds and targets are completely different (case - 2)
print(mcc(torch.ones(10), torch.zeros(10)))
```

```python
tensor(0)
tensor(0)
tensor(0)
```
In this [blog](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6413-7), They are handling these cases slightly different. Below are the MCC values using their [implementation](https://gist.github.com/lokesh1199/3d899b03e37107565cd91d6ad6d8db90). These values represent model performance more accurately than the torchmetrics version.

case - 1 -> MCC = 1.0
case - 2 -> MCC = -1.0

Kindly tell if this should be handled differently?
</div>

**Comment by SkafteNicki** [comment]
Hi @lokesh1199, thanks for raising this issue.
To be fair we have just converted the implementation from sklearn, so I have never though too long about the actual code.
Looking through their issues I can see someone has raised a similar point to what you are doing here:
https://github.com/scikit-learn/scikit-learn/issues/25258
That said, after looking at the paper I can see two arguments:
* If we just look at the formula, the score is undefined (division by 0) for the cases where everything matches and it is therefore a completely arbitrary what the metric should return in this case
* Based on the analysis from the paper if we look at the limit, it does make sense that it should return -1/+1 in the corner cases

I therefore think it makes sense to change the implementation, just for the sake of result making better sense.

**Comment by lokesh1199** [comment]
Hi @SkafteNicki,

The formula which was suggested in the [paper](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6413-7) might be wrong (#1812)

### Number of false negatives / MCC (torchmetrics)
```python
n = 1000
mcc_scores = []
targets = torch.Tensor([1] * n)
for i in range(n+1):
    preds = torch.Tensor([1] * (n - i) + [0] * i)
    mcc_scores.append(mcc_torch(preds, targets))
plt.ylim(-1, 1)
plt.plot(range(n+1), mcc_scores)
```
![image](https://github.com/Lightning-AI/torchmetrics/assets/68476180/90d81d60-7042-413f-848f-f74af8a2501e)
As we can see, even for the worst case, the metric is not reaching to -1

### Number of false positives / MCC (torchmetrics)
```python
n = 1000
mcc_scores = []
targets = torch.Tensor([0] * n)
for i in range(n+1):
    preds = torch.Tensor([0] * (n - i) + [1] * i)
    mcc_scores.append(mcc_torch(preds, targets))
plt.ylim(-1, 1)
plt.plot(range(n+1), mcc_scores)
```
![image](https://github.com/Lightning-AI/torchmetrics/assets/68476180/1be299bc-782c-4c71-947a-e2bad8dc9ab3)
MCC metric should be same even if we switch the positive and negative case. But as we can clearly see above, If the input contains all `True negatives`, the MCC is -1.

### Fix
This will fix the above mentioned issues.
```python
def mcc_eps(pred, target):
    conf_mat = binary_confusion_matrix(pred, target)

    tn, fp, fn, tp = conf_mat.reshape(-1)

    return _mcc(tp,tn, fp, fn)

def _mcc(tp, tn, fp, fn):

    if tp + tn != 0 and fp + fn == 0:
        return 1

    if tp + tn == 0 and fp + fn != 0:
        return -1


    numerator = tp*tn - fp*fn
    denom = (tp + fp) * (tp + fn) * (tn + fp) * (tn + fn)


    if denom == 0:
        if tp == 0 or tn == 0:
            a = tp + tn
        
        if fp == 0 or fn == 0:
            b = fp + fn


        eps = torch.finfo(torch.float32).eps

        numerator = math.sqrt(eps)  * (a - b)
        denom = (tp + fp + eps) * (tp + fn + eps) * (tn + fp + eps) * (tn + fn + eps)

    return numerator / math.sqrt(denom)
```
Plots after the above fix,
### False negatives
![image](https://github.com/Lightning-AI/torchmetrics/assets/68476180/92590f0e-7cce-4943-baf5-077d66a878af)


### False positves
![image](https://github.com/Lightning-AI/torchmetrics/assets/68476180/74531ea8-14fd-44ac-97e2-f4ed6f8c7dd9)
